sdk/python/feast/pyspark/launchers/aws/emr_utils.py

```diff
@@ -0,0 +1,363 @@
+import hashlib
+import json
+import logging
+import os
+import random
+import string
+import tempfile
+import time
+from collections import defaultdict
+from io import BytesIO
+from pathlib import Path
+from typing import IO, Any, Dict, List, NamedTuple, Optional, Tuple
+
+import boto3
+import botocore
+import pandas
+import yaml
+
+from feast.client import Client
+from feast.feature_table import FeatureTable
+from feast.value_type import ValueType
+
+log = logging.getLogger("aws")
+
+SUPPORTED_EMR_VERSION = "emr-6.0.0"
+STREAM_TO_ONLINE_JOB_TYPE = "STREAM_TO_ONLINE_JOB"
+OFFLINE_TO_ONLINE_JOB_TYPE = "OFFLINE_TO_ONLINE_JOB"
+HISTORICAL_RETRIEVAL_JOB_TYPE = "HISTORICAL_RETRIEVAL_JOB"
+
+
+# Mapping of EMR states to "active" vs terminated for whatever reason
+ACTIVE_STEP_STATES = ["PENDING", "CANCEL_PENDING", "RUNNING"]
+TERMINAL_STEP_STATES = ["COMPLETED", "CANCELLED", "FAILED", "INTERRUPTED"]
+
+# Mapping of EMR states to generic states
+IN_PROGRESS_STEP_STATES = ["PENDING", "CANCEL_PENDING", "RUNNING"]
+SUCCEEDED_STEP_STATES = ["COMPLETED"]
+FAILED_STEP_STATES = ["CANCELLED", "FAILED", "INTERRUPTED"]
+
+
+def _sanity_check_cluster_template(template: Dict[str, Any], template_path: str):
+    """
+    Sanity check the run job flow template. We don't really have to do this here but if the spark
+    job fails you'll only find out much later and this is annoying. Those are not exhaustive, just
+    some checks to help debugging common configuration issues.
+    """
+
+    releaseLabel = template.get("ReleaseLabel")
+    if releaseLabel != SUPPORTED_EMR_VERSION:
+        log.warn(
+            f"{template_path}: ReleaseLabel is set to {releaseLabel}. Recommended: {SUPPORTED_EMR_VERSION}"
+        )
+
+
+def _load_new_cluster_template(path: str) -> Dict[str, Any]:
+    with open(path) as f:
+        template = yaml.safe_load(f)
+        _sanity_check_cluster_template(template, path)
+        return template
+
+
+def _random_string(length) -> str:
+    return "".join(random.choice(string.ascii_lowercase) for _ in range(length))
+
+
+def _s3_split_path(path: str) -> Tuple[str, str]:
+    """ Convert s3:// url to (bucket, key) """
+    assert path.startswith("s3://")
+    _, _, bucket, key = path.split("/", 3)
+    return bucket, key
+
+
+def _hash_fileobj(fileobj: IO[bytes]) -> str:
+    """ Compute sha256 hash of a file. File pointer will be reset to 0 on return. """
+    fileobj.seek(0)
+    h = hashlib.sha256()
+    for block in iter(lambda: fileobj.read(2 ** 20), b""):
+        h.update(block)
+    fileobj.seek(0)
+    return h.hexdigest()
+
+
+def _s3_upload(
```

Comment:
Can this be moved to staging client?

Reply:
Yes, if it is ok i'd do it in a separate PR though. It is not exactly 1-1 with the current staging client interface.

Reply:
I submitted a PR here: https://github.com/feast-dev/feast/pull/1063/files